src/context/values.js

```diff
@@ -0,0 +1,720 @@
+    const refreshParent = used.pending == Pending.PENDING_REFRESH_PARENT;
+
+    let newValue;
+    try {
+      newValue = this.calc_(used, refreshParent);
+    } catch (e) {
+      // This is the narrowest catch to avoid unrelated values breaking each
+      // other. The only exposure to the user-code are `needsParent` and
+      // `compute` methods in the `ContextProp`.
+      rethrowAsync(e);
+    }
+
+    // Reset pending flag. It's good to reset it after the calculation to
+    // ensure that deps are automatically covered.
+    used.pending = Pending.NOT_PENDING;
+
+    // Check if the value has been updated.
+    this.maybeUpdated_(used, newValue);
+  }
+
+  /**
+   * @param {!UsedDef} used
+   * @param {*} value
+   * @private
+   */
+  maybeUpdated_(used, value) {
+    const {prop, value: oldValue} = used;
+    const {key} = prop;
+    const usedByKey = this.usedByKey_;
+    if (
+      oldValue === value ||
+      used !== (usedByKey && usedByKey.get(key)) ||
+      !this.isConnected_()
+    ) {
+      // Either the value didn't change, or no one needs this value anymore.
+      return;
+    }
+
+    used.value = value;
+
+    // Notify subscribers.
+    const {subscribers} = used;
+    subscribers.forEach((handler) => {
+      handler(value);
+    });
+  }
+
+  /**
+   * The used value calculation algorithm.
+   *
+   * @param {!UsedDef} used
+   * @param {boolean} refreshParent
+   * @return {*|undefined} The used value.
+   * @private
+   */
+  calc_(used, refreshParent) {
+    devAssert(this.isConnected_());
+
+    const {prop, depValues} = used;
+    const {key, compute, defaultValue} = prop;
+
+    const inputsByKey = this.inputsByKey_;
+    const inputs = inputsByKey && inputsByKey.get(key);
+    const inputValues = inputs && inputs.values;
+
+    // Calculate parent value.
+    const needsParent = calcNeedsParent(prop, inputValues);
+
+    // Refresh parent if requested.
+    if (refreshParent || needsParent != Boolean(used.parentValueNode)) {
+      const newParentValueNode = needsParent
+        ? findParent(this.contextNode_, hasInput, prop, /* includeSelf */ false)
+        : null;
+      this.updateParentValueNode_(used, newParentValueNode);
+    }
+
+    // If no parent node is found, use the default value.
+    const parentValue = defined(used.parentValue)
+      ? used.parentValue
+      : needsParent && !used.parentValueNode && inputValues
+      ? defaultValue
+      : undefined;
+
+    // Calculate the "used" value.
+    let newValue = undefined;
+    const ready =
+      depValues.every(defined) && (!needsParent || defined(parentValue));
+    if (ready) {
+      const {node} = this.contextNode_;
+      if (inputValues && !compute) {
+        newValue = inputValues[0];
+      } else if (isRecursive(prop)) {
+        if (inputValues) {
+          // The node specifies its own input values and they need to be
+          // recomputed with parent and dep values.
+          newValue = compute(
```

Comment:
I think we need to check if there's a `compute`?

Reply:
The previous `if` is `if (inputValues && !compute)`, so it's already covered.